### ISSUE TITLE
Get programmatically a list of used plugins

### ISSUE THREAD

**Issue by aight8** [babel/babel#5374]
# Feature request

Sadly there is currently no way to get a list of the used plugins (**inclusively which are added by presets**). The problem is that for debugging reasons I want to write a checker for the **webpack** package which checks if some module transformer are used.

Also optionally it should output the whole plugin list (for example env/latest etc. are dynamical on time) - but for this I must access it programatically!

**Does the babel team plan to implement a functionality such this to the babel core?**

Please help!

**Comment by loganfsmyth** [comment]
Part of the issue here is that I don't think there is a clear list of plugins to give. If a user for instance does

    plugins: [require('custom-plugin')]

all Babel knows is that it was passed a plugin, it doesn't need to have a name, and we have no way of knowing where it came from.

We do have https://github.com/babel/babel/issues/2960 which I think is roughly what you are asking for?

**Comment by aight8** [comment]
You have right. Then let's make a strategy change. Let just name the plugins internally, they have dozens of props which, can we just add a **name** prop to it? The most of the plugins are managed by babel luckily and it wouldn't BC-break anything.

just by the way: What happen if you add a plugin multiple times (for example by different presets)? Does the second plugin instance run but find nothing to transform?

**Comment by hzoo** [comment]
You can pass a name to plugins now - from http://astexplorer.net/

```
export default function (babel) {
  const { types: t } = babel;
  
  return {
    name: "ast-transform", // not required
    visitor: {
      Identifier(path) {}
    }
  };
}
```

**Comment by aight8** [comment]
@hzoo That's awesome - I must search that PR. All the current babel plugin's become a name also ? ;)

FYI: With this naming webpack could check if babel do any module transformation and prevent webpack by tree-shaking. I'm working on a PR there.

**Comment by hzoo** [comment]
I didn't know about it either earlier but it seems like it was always there? I just don't think the regular plugins have a name. I believe what @kittens said before was that if you pass a string it uses that as the name, otherwise it doesn't (require)

**Comment by aight8** [comment]
Yeah than it's optional. But wouldn't be it nice to declare this name prop for all babel regular plugins?

**Comment by loganfsmyth** [comment]
The other part of my concern here is that the implication here is that third-party tools would be dependent on the names of our plugins seems terrible. If someone wants to come along and implement their own plugin or fork one of our own, they should be able to do so without having to worry about the fact that the plugin name would be different.

I'm up for exposing the name for debugging potentially, but I really don't think it should be a production feature.

Can you elaborate on what you are trying to do for Webpack?

**Comment by aight8** [comment]
Hm okey I wondering about this argument, I guess over 99% don't fork babel plugins.
If someone fork a babel plugin or implement their own, everything is working fine. Babel must not limiting named plugins to one instance. It's for debugging purposes only.

I currently work on a webpack related package which analyse the webpack config and hint problems and improvements to the user.

**Comment by loganfsmyth** [comment]
Could you make a Webpack plugin that looked at the loader input and output and if it saw conversions from ES6 modules to CommonJS from before to after the loader, complain? If you focus on the input and output, you'd avoid all the details about Babel.

**Comment by aight8** [comment]
Hm, this becomes big complexity then especially it should analyse the configuration-only without processing webpack. Also use the webpack object constructor to check if the basic structure is okey and use some libs which are required.

The list of checks are maintained and growing continuously - it's like a linter/analyser. However there are a handful (some dozens) of external loaders/plugins which are used primarily. Babel is one of the most popular and affects the tree-shaking which is also one of the biggest optimisation.

So I thought it's an easy-fix to just ask the used babel-loader "hey, do you use one of these plugins transform-es2015-modules-*?" - which are imported mostly by one of the dozens of custom presets out there. A 3 liner for literally 99.9% of the use cases.

This would be enough to help a lot of webpack users but I must solve that problem somehow else - currently don't know how but I want to prevent much complexity.

**Comment by aight8** [comment]
I get rid of that, using babel-core and compile a small snippet with and without the provided babel config and than just look after import/export node type if they are got replaced. An improvement were to check which kind of modules are used - that just looks cool.

However really useful were to output **which** plugin (and from which preset) affected that.
Also, beside that, it's always informative as overview when users can see the used plugins (maybe even as tree) - this is then exactly this issue #2960 (or this WIP PR https://github.com/babel/babel/pull/4806)

I think also for those issue it were enough to save the *loc* to the Plugin object, then you can leave it to the dev how they use the path to show a plugin name.

**Comment by nicolo-ribaudo** [comment]
Is this fixed by `babel.loadOptions` (https://babeljs.io/docs/en/babel-core#loadoptions)? cc @loganfsmyth

**Comment by 0xdevalias** [comment]
> Is this fixed by `babel.loadOptions` (https://babeljs.io/docs/en/babel-core#loadoptions)? cc @loganfsmyth

It would be nice to get this exposed in the CLI as well, as per https://github.com/babel/babel/pull/4806
